add rdkit_2d featurizer

KnathanM opened the pull request:
## Description
People want the rdkit_2d_normalized featurizer available from the commandline, probably because it was helpful in v1 (the [docs](https://chemprop.readthedocs.io/en/latest/tutorial.html#molecule-level-rdkit-2d-features) say "As a starting point, we recommend using pre-normalized RDKit features by using the --features_generator rdkit_2d_normalized --no_features_scaling flags"

This PR doesn't do that but is a step in that direction. 

This PR implements a featurizer that makes the rdkit 2d features. But they aren't normalized, as that was done in Descriptastorus with precomputed scaling ([here](https://github.com/bp-kelley/descriptastorus/blob/master/descriptastorus/descriptors/dists.py)). Also Descriptastorus leaves out these descriptors : `['AvgIpc', 'BCUT2D_CHGHI', 'BCUT2D_CHGLO', 'BCUT2D_LOGPHI', 'BCUT2D_LOGPLOW', 'BCUT2D_MRHI', 'BCUT2D_MRLOW', 'BCUT2D_MWHI', 'BCUT2D_MWLOW', 'SPS']`. At least "SPS" was added more recently. I haven't checked the others. 

Also note that "SPS" divides by the number of heavy atoms, so the featurizer has to check for that. 

## Questions
Would we prefer to just make Descriptastorus a required dependency again?

## Checklist
- [ ] (if appropriate) unit tests added?
Tests not yet added

shihchengli reviewed (commented):
I think it would be better not to add Descriptastorus as a required dependency if it is not too difficult to write our own versions and we accept additional lines of code for the implementation of [RDKit2DNormalized](https://github.com/bp-kelley/descriptastorus/blob/master/descriptastorus/descriptors/rdNormalizedDescriptors.py#L79).

hwpang commented on chemprop/featurizers/molecule.py:
```diff
@@ -41,3 +41,24 @@ class MorganBinaryFeaturizer(MorganFeaturizerMixin, BinaryFeaturizerMixin, Vecto
+        else:
+            features = np.array([func(mol) for name, func in sorted(Descriptors.descList)])
+
+        NAN_TOKEN = 0
+        features[np.isnan(self.x_d)] = NAN_TOKEN
```
self.x_d is molecular descriptor for a data point

shihchengli replied:
I see, but if it is an attribute in the data points, can it be accessed here? Besides, it seems this code [here](https://github.com/chemprop/chemprop/blob/main/chemprop/data/datapoints.py#L43-L45) already handled the issues about nan values.

KnathanM replied:
I had copied this code (but accidentally left self.x_d) because some of the rdkit2d features would return nan (https://github.com/rdkit/rdkit/discussions/5824) and I feel it is the responsibility of the featurizer to not return nans. 

But in any event, these features that would return nan are not calculated by descriptastorus. There are 12 features that rdkit descriptors have that Descriptastorus doesn't. But I figure the consistency with chemprop v1 is worth not having these extra features especially since a user can still calculate them manually.

KnathanM commented:
I have changed the implementation to rely on Descriptastorus. This is what was done in v1. 

I will also note that the featurizer works for mols without any heavy atoms
```
from chemprop.featurizers.molecule import RDKit2DFeaturizer
from rdkit import Chem

mol = Chem.MolFromSmiles('[H][H].[H]')
featurizer = RDKit2DFeaturizer()
featurizer(mol)
```

hwpang commented:
Thanks for adding the new featurizer! Can you add CLI test for it?

JacksonBurns reviewed (changes requested):
The conda recipe will only work with packages on `conda-forge`, so we cannot use the git-based install.

JacksonBurns commented:
Elaborating on my comment above:
 - this PR has only added the dependency to the `pyproject.toml`, meaning that the `environment.yml`-based tests fail because it is missing
 - it can be added to `environment.yml` but not the conda recipe we use to distribute on `conda-forge` (meaning we would not have `conda install chemprop` anymore)
 - I think we should avoid adding `descriptastorus` as a required dependency, since the repository is rarely updated and has broken on us in the past

My suggested alternatives:
 - vendor just the one function we want
 - add our own feature calculator and standardizer

There's also a theoretical question about whether or not using the `descriptastorus` features is a data leak. My thinking is that the distributions used to rescale the features are parameterized based on some set that I don't know (does anyone?) that could possibly contain overlapping molecules with a testing or validation set, which would shift the distribution of said features to look more like training and thus constitute a data leak.

KnathanM commented:
Ready for another review.

I removed descriptastorus as a dependency (see below for why). 

>  Can you add unit test and/or CLI test for it?

I added a clunky test, open to improvements on it. Mostly I wanted to make sure that the molecule featurizers don't change how they featurize (e.g. RDKit adds another molecule feature to its list), because Chemprop models from the CLI rely on the featurizers not changing. 

I also added a note about using descriptastorus to calculate normalized features to the docs. 

### why remove descriptastorus

@JacksonBurns had some thoughts about this:

> it can be added to environment.yml but not the conda recipe we use to distribute on conda-forge (meaning we would not have conda install chemprop anymore)

Turns out descriptastorus is on conda-forge: https://anaconda.org/conda-forge/descriptastorus

> I think we should avoid adding descriptastorus as a required dependency, since the repository is rarely updated and has broken on us in the past

I also think that the previous trouble we had with descriptastorus wasn't that bad. The latest version has an error that makes it incompatible with scipy<1.9, but that version of scipy was released two years ago. The simple solution was to update scipy. This error was fixed two months ago on GitHub, but hasn't been released as a new version. 

> There's also a theoretical question about whether or not using the descriptastorus features is a data leak. My thinking is that the distributions used to rescale the features are parameterized based on some set that I don't know (does anyone?) 

I think this is something to consider. The main reason I chose to not have descriptastorus available from the CLI is that I don't know where their scaling came from. Including it in the CLI is essentially a recommendation to use it as it only requires specifying a flag, which I don't feel I can recommend using if I don't know where it came from. Mostly I want to encourage chemprop users to be aware of what information their model is actually using to make predictions. If they are using descriptors, they should hopefully have some idea of what those descriptors are. 

A secondary reason I chose to not use descriptastorus is that it is missing a few of the newer molecular features added to RDKit. See the linked issues for details.

JacksonBurns commented:
Didn't know it was on `conda-forge`! That changes my opinion, and I would be OK with it from a software perspective.

I'll note too that the incompatible version of `scipy` cannot even be installed on the versions of Python that Chemprop v2 supports, see: https://docs.scipy.org/doc/scipy/dev/toolchain.html#numpy

Someone else has also just opened an issue asking about the scaling on the `descriptastorus` repo: https://github.com/bp-kelley/descriptastorus/issues/31 I'm hoping for a satisfactory answer.

JacksonBurns commented:
I've started a small test PR on the conda recipe to see if the build will work with `descriptastorus`.

JacksonBurns commented on chemprop/featurizers/molecule.py:
```diff
@@ -41,3 +43,28 @@ class MorganBinaryFeaturizer(MorganFeaturizerMixin, BinaryFeaturizerMixin, Vecto
+        else:
+            features = np.array([func(mol) for name, func in Descriptors.descList], float)
+
+        NAN_TOKEN = 0
```
Not sure I see the value in setting this here and then using it immediately versus just adding a comment that says 'set np.nan to zero'

KnathanM replied:
The main value is that it is consistent with how we deal with nans in the extra features/descriptors the user gives when making a data point (https://github.com/chemprop/chemprop/blob/06171ace741b1641deddf978a96d99a68236c352/chemprop/data/datapoints.py#L35). That said, we already take care of this in datapoint, so probably I can remove this nan_to_zero conversion and return an array that potentially has some nans. This is different than what I previously thought, that the featurizers were responsible for not returning nans. Currently, I'm of the opinion we shouldn't hide the nans from the user, except when we need to like when making the datapoints because the model can't take nans.

JacksonBurns commented on chemprop/featurizers/molecule.py:
```diff
@@ -41,3 +43,28 @@ class MorganBinaryFeaturizer(MorganFeaturizerMixin, BinaryFeaturizerMixin, Vecto
+        return len(Descriptors.descList)
+
+    def __call__(self, mol: Chem.Mol) -> np.ndarray:
+        if mol.GetNumHeavyAtoms() == 0:
+            features = np.array(
+                [0 if name == "SPS" else func(mol) for name, func in Descriptors.descList], float
+            )
+        else:
+            features = np.array([func(mol) for name, func in Descriptors.descList], float)
```
```suggestion
        features = np.array(
            [func(mol) for _, func in filter(lambda i: mol.GetNumHeavyAtoms() > 0 or i[0] != "SPS", Descriptors.descList)],
            dtype=float,
        )
```

I think this is the same and avoids duplicating a line

KnathanM replied:
Thank you for writing that with `filter`. I switched am putting `i[0] != "SPS"` first so that we don't need to call `GetNumHeavyAtoms` most of the time, only when the function name is SPS.

JacksonBurns commented on docs/source/tutorial/cli/train.rst:
```diff
@@ -251,22 +251,41 @@ Note that bond descriptors are not currently supported because the post message
 ^^^^^^^^^^^^^^^^^^^^^^^^^^
 
-Morgan fingerprints can be generated as molecular 2D features using :code:`--features-generators`:
+Chemprop provides several molecule featurizers that automatically calculates molecular features and uses them as extra datapoint descriptors. These are specified using :code:`--features-generators` followed by one or more of the following:
```
```suggestion
Chemprop provides several molecule featurizers that automatically calculate molecular features and uses them as extra datapoint descriptors. These are specified using :code:`--features-generators` followed by one or more of the following:
```

JacksonBurns commented on tests/cli/test_cli_regression_mol.py:
```diff
@@ -431,7 +431,7 @@ def test_hyperopt_quick(monkeypatch, data_path, tmp_path):
         "hyperopt",
         "--features-generators",
-        "morgan_count",
+        "morgan_binary",
```
Are we no longer testing `morgan_count`?

KnathanM replied:
Before we tested `morgan_count` multiple times. `morgan_count` is still tested in `test_optuna_quick`.

shihchengli replied:
As long as we have good unit tests for each molecular featurizer and can ensure that the training and hpopt work with `--features-generators`, I think it is fine. I can understand it would be great to cover all possible cases to avoid unexpected errors in the future. For example, we should include a CLI test with data points whose RDKit features might be NaN so we can test that the data point can successfully convert NaN to 0. The drawback is that if we consider every possible case, it would likely take too much time for the checks.

shihchengli commented on chemprop/featurizers/molecule.py:
```diff
@@ -9,7 +11,7 @@
 
 
-class MorganFeaturizerMixin:
+class MorganFeaturizerMixin(VectorFeaturizer[Mol]):
```
The `MorganFeaturizerMixin` lacks a `__call__` function to convert input `Mol`s into output vectors. Therefore, I think the original version is ok.

KnathanM replied:
Good point

shihchengli commented on docs/source/tutorial/python/featurizers/molecule_featurizers.ipynb:
```diff
@@ -1,184 +1,300 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "If you only have molecule features for one molecule per datapoint, those features can be used directly as extra datapoint descriptures. If you have multiple molecules with extra features, or other extra datapoint descriptors, they first need to be concatenated into a single numpy array."
```
```suggestion
    "If you only have molecule features for one molecule per datapoint, those features can be used directly as extra datapoint descriptors. If you have multiple molecules with extra features, or other extra datapoint descriptors, they first need to be concatenated into a single numpy array."
```

shihchengli replied:
I think this hasn't been addressed yet.

StefanIsSmart commented:
Hello, anyone know the answer about  https://github.com/bp-kelley/descriptastorus/issues/31 ?

JacksonBurns commented:
> Hello, anyone know the answer about  https://github.com/bp-kelley/descriptastorus/issues/31 ?

No, we are also curious.

shihchengli commented on chemprop/featurizers/molecule.py:
```diff
@@ -41,3 +49,68 @@ class MorganBinaryFeaturizer(MorganFeaturizerMixin, BinaryFeaturizerMixin, Vecto
+
+if not NO_DESCRIPTASTORUS:
+
+    class V1RDKit2DFeaturizerMixin(VectorFeaturizer[Mol]):
```
What is the difference between `RDKit2DFeaturizer` and `V1RDKit2DFeaturizer`? Should we just keep one of them to avoid confusion?

shihchengli replied:
It seems the new one has 10 more descriptors. Maybe we can just keep one of them.
> 'AvgIpc', 'BCUT2D_CHGHI', 'BCUT2D_CHGLO', 'BCUT2D_LOGPHI', 'BCUT2D_LOGPLOW', 'BCUT2D_MRHI', 'BCUT2D_MRLOW', 'BCUT2D_MWHI', 'BCUT2D_MWLOW', and 'SPS'

KnathanM replied:
If we had to keep just one, I would recommend it being the V1RDKit2DFeaturizer because if someone is trying to use a v1 model that used rdkit features, they would need that featurizer. 

I added a new version because of the extra descriptors that have been added since descriptastorus. It could be nice to let people get these more descriptors. My intention is that all new models would use RDKit2DFeaturizer and that the V1 version would only be used for models that were trained in v1. 

Let me know if you feel strongly, or we can bring it up in the next meeting.

shihchengli replied:
I am okay with keeping both of them. However, in the long term, I don't think people will care about V1RDKit2DFeaturizer anymore, as it will become a legacy we used in the past. If someone wants to do a benchmark with this smaller set, they should take responsibility for generating those features correctly via a notebook and provide them as additional features.

shihchengli commented on chemprop/featurizers/molecule.py:
```diff
@@ -41,3 +49,68 @@ class MorganBinaryFeaturizer(MorganFeaturizerMixin, BinaryFeaturizerMixin, Vecto
+    class V1RDKit2DFeaturizer:
+        """Mock implementation raising an ImportError if descriptastorus cannot be imported."""
+
+        def __init__(self):
+            raise ImportError(
+                "Failed to import descriptastorus. Please install descriptastorus "
+                "(https://github.com/bp-kelley/descriptastorus) to use RDKit 2D features."
+            )
```
I saw that this is a mock implementation. You might already have an idea of how to implement it in a real implementation. Here is my two cents:
Move this into `V1RDKit2DFeaturizerMixin` and check if `NO_DESCRIPTASTORUS` is True.

KnathanM replied:
Good idea

shihchengli commented on tests/unit/featurizers/test_molecule.py:
```diff
@@ -0,0 +1,191 @@
+# flake8: noqa
```
Also add a test for `V1RDKit2DNormalizedFeaturizer`?

shihchengli commented:
It seems we are going to add Descriptastorus as a dependency. Could you modify the pyproject.toml and environment.yml? accordingly?

JacksonBurns commented:
> I've started a small test PR on the conda recipe to see if the build will work with `descriptastorus`.

The linked PR on the conda recipe confirmed that the build will work.

KnathanM commented:
@shihchengli Am I missing any changes you'd like to see in this PR before merging?

shihchengli commented on chemprop/featurizers/molecule.py:
```diff
@@ -41,3 +49,62 @@ class MorganBinaryFeaturizer(MorganFeaturizerMixin, BinaryFeaturizerMixin, Vecto
+    def __call__(self, mol: Chem.Mol) -> np.ndarray:
+        features = np.array(
+            [
+                func(mol)
+                for _, func in filter(
+                    lambda i: i[0] != "SPS" or mol.GetNumHeavyAtoms() > 0, Descriptors.descList
```
```suggestion
            0.0 if name == "SPS" and mol.GetNumHeavyAtoms() == 0 else func(mol)
            for name, func in Descriptors.descList
```
For species without heavy atoms, the length of RDKit2D will be shorter by 1. I think it should be padded with 0.

KnathanM replied:
Thanks for catching that!

shihchengli commented on docs/source/tutorial/cli/train.rst:
```diff
@@ -248,22 +248,32 @@ Note that bond descriptors are not currently supported because the post message
-Morgan fingerprints can be generated as molecular 2D features using :code:`--molecule-featurizers`:
+Chemprop provides several molecule featurizers that automatically calculate molecular features and uses them as extra datapoint descriptors. These are specified using :code:`--molecule-featurizers` followed by one or more of the following:
 
 * :code:`morgan_binary` binary Morgan fingerprints, radius 2 and 2048 bits.
 * :code:`morgan_count` count-based Morgan, radius 2 and 2048 bits.
```
```suggestion
* :code:`morgan_binary` binary Morgan fingerprints, radius 2 and 2048 bits
* :code:`morgan_count` count-based Morgan, radius 2 and 2048 bits
```

shihchengli commented on docs/source/tutorial/cli/train.rst:
```diff
@@ -248,22 +248,32 @@ Note that bond descriptors are not currently supported because the post message
-Morgan fingerprints can be generated as molecular 2D features using :code:`--molecule-featurizers`:
+Chemprop provides several molecule featurizers that automatically calculate molecular features and uses them as extra datapoint descriptors. These are specified using :code:`--molecule-featurizers` followed by one or more of the following:
 
 * :code:`morgan_binary` binary Morgan fingerprints, radius 2 and 2048 bits.
 * :code:`morgan_count` count-based Morgan, radius 2 and 2048 bits.
+* :code:`rdkit_2d` RDKit 2D features
+* :code:`v1_rdkit_2d` The RDKit 2D features used in Chemprop v1
+* :code:`v1_rdkit_2d_normalized` The normalized RDKit 2D features used in Chemprop v1
```
```suggestion
 * :code:`morgan_binary` binary Morgan fingerprints, radius 2 and 2048 bits
 * :code:`morgan_count` count-based Morgan, radius 2 and 2048 bits
 * :code:`rdkit_2d` RDKit 2D features
 * :code:`v1_rdkit_2d` The RDKit 2D features used in Chemprop v1
 * :code:`v1_rdkit_2d_normalized` The normalized RDKit 2D features used in Chemprop v1
```

shihchengli commented on docs/source/tutorial/cli/train.rst:
```diff
@@ -248,22 +248,32 @@ Note that bond descriptors are not currently supported because the post message
+
+   The RDKit 2D features are not normalized. The :code:`StandardScaler` used in the CLI to normalize is non-optimal for some of the RDKit features. It is recommended to precompute and scale these features outside of the CLI using an appropriate scaler and then provide them using :code:`--descriptors-path` and :code:`--no-descriptor-scaling` as described above. 
+
+   In Chemprop v1, :code:`descriptastorus`` was used to calculate RDKit 2D features. This package offers normalization of the features, with the normalizations fit to a set of molecules randomly selected from ChEMBL. Several descriptors have been added to :code:`rdkit` recently which are not included in :code:`descriptastorus` including 'AvgIpc', 'BCUT2D_CHGHI', 'BCUT2D_CHGLO', 'BCUT2D_LOGPHI', ' ...
```
```suggestion
   In Chemprop v1, :code:`descriptastorus` was used to calculate RDKit 2D features. This package offers normalization of the features, with the normalizations fit to a set of molecules randomly selected from ChEMBL. Several descriptors have been added to :code:`rdkit` recently which are not included in :code:`descriptastorus` including 'AvgIpc', 'BCUT2D_CHGHI', 'BCUT2D_CHGLO', 'BCUT2D_LOGPHI', 'BCUT2D_LOGPLOW', 'BCUT2D_MRHI', 'BCUT2D_MRLOW', 'BCUT2D_MWHI', 'BCUT2D_MWLOW', and 'SPS'.
```

shihchengli commented on docs/source/tutorial/python/featurizers/molecule_featurizers.ipynb:
```diff
@@ -118,13 +124,160 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The rdkit featurizers from v1 are also available. They rely on the `descriptastorus` package which can be found at `https://github.com/bp-kelley/descriptastorus`. This package doesn't include the following rdkit descriptors: `['AvgIpc', 'BCUT2D_CHGHI', 'BCUT2D_CHGLO', 'BCUT2D_LOGPHI', 'BCUT2D_LOGPLOW', 'BCUT2D_MRHI', 'BCUT2D_MRLOW', 'BCUT2D_MWHI', 'BCUT2D_MWLOW', 'SPS']`. Scaled versions of  ...
```
```suggestion
    "The rdkit featurizers from v1 are also available. They rely on the `descriptastorus` package which can be found at [https://github.com/bp-kelley/descriptastorus](https://github.com/bp-kelley/descriptastorus). This package doesn't include the following rdkit descriptors: `['AvgIpc', 'BCUT2D_CHGHI', 'BCUT2D_CHGLO', 'BCUT2D_LOGPHI', 'BCUT2D_LOGPLOW', 'BCUT2D_MRHI', 'BCUT2D_MRLOW', 'BCUT2D_MWHI', 'BCUT2D_MWLOW', 'SPS']`. Scaled versions of these descriptors are available, though it is unknown which molecules were used to fit the scaling, so this may be a dataleak depending on the test set used to evaluate model performace. See this [issue](https://github.com/bp-kelley/descriptastorus/issues/31) for more details about the scaling. "
```

shihchengli commented on docs/source/tutorial/python/featurizers/molecule_featurizers.ipynb:
```diff
@@ -118,13 +124,160 @@
+    }
+   ],
+   "source": [
+    "molecule_featurizer = V1RDKit2DFeaturizer()\n",
+    "molecule_featurizer = V1RDKit2DNormalizedFeaturizer()\n",
```
Is the first line intended to demonstrate the availability of this featurizer? Adding another cell to show the difference might be overkill. I'm fine with keeping it as it is now.

KnathanM replied:
Yes it is intended to demonstrate the availability

shihchengli reviewed (commented):
Thanks for improving this PR along the way. These are my final comments on this PR. After considering and addressing them, I would be happy to merge it.

JacksonBurns reviewed (commented):
This looks great and is ready to merge, I just want to clear something up for future development purposes.

One overall question - we allow for the import of `descriptastorus` to fail (i.e. allowing for it to not be installed) but we have it in the PyPI and Conda recipe (i.e. it will always be installed). Why? My initial thinking is that we would want to not make it required or not allow it to be missing.

KnathanM commented:
That's a good point. I think we are okay requiring descriptastorus because it available on PyPI and Conda. Let's see if the tests work when it is required. Current users will need to install descriptastorus themselves when they pull in these changes, but that shouldn't be too hard?

JacksonBurns commented:
Agreed, should be straightforward. Approving!
